### PR TITLE
Fix overflow in torch.linalg.norm / vector_norm for float32/float64 (fixes #193006)

### LINKED ISSUE #193006: `torch.linalg.norm` returns `inf` for norms well inside the dtype range, and its gradient is silently `0` (float32/float64)
### 🐛 Describe the bug

`torch.linalg.norm` computes `sqrt(sum(x**2))` directly, so whether the call overflows is
decided by the intermediate **sum of squares**, not by `‖x‖`. A vector whose norm is `1e20`
in float32 — five orders of magnitude below the float32 max of `3.4e38` — returns `inf`,
because `sum(x**2)` is `1e40`.

The gradient is the more damaging half. `backward()` evaluates `x / ‖x‖` as `x / inf` and
produces **`0.0` for every entry**, with no `inf` or `nan` anywhere in the gradient to
signal that anything went wrong. `torch.nn.functional.normalize`, which is built on the
same reduction, likewise returns an all-zero vector rather than a unit vector.

`scipy.linalg.norm` returns the correct value for the identical input, so this is not an
inherent limit of the dtype.

## Reproducer

```python
import torch, numpy as np, scipy.linalg

# true ||v|| = 1e20, well inside float32 range (max 3.4e38).
# Only the intermediate sum(v^2) = 1e40 overflows.
v = np.array([1e20, 1.0, 1.0], dtype=np.float32)
x = torch.tensor(v, requires_grad=True)
y = torch.linalg.norm(x); y.backward()

print("true ||v||        : 1e+20      grad [1.0, 1e-20, 1e-20]")
print("torch.linalg.norm :", float(y))
print("torch grad        :", x.grad.numpy())
print("scipy.linalg.norm :", scipy.linalg.norm(v))
```

### Actual output

```
true ||v||        : 1e+20      grad [1.0, 1e-20, 1e-20]
torch.linalg.norm : inf
torch grad        : [0. 0. 0.]
scipy.linalg.norm : 1.0000000200408773e+20
```

## Affected entry points

Nine float32 entries of `1e19`, as a vector and as a 3×3 (true norm `3e19` in both cases):

```python
v = torch.full((9,),  1e19, dtype=torch.float32)
m = torch.full((3, 3), 1e19, dtype=torch.float32)
```

```
torch.linalg.vector_norm(v)        -> inf
torch.linalg.matrix_norm(m)        -> inf
torch.nn.functional.normalize(v)   -> 0.0   (expected 0.3333)
```

`torch.linalg.norm` and `torch.norm` dispatch to the two primitives above, so they return
`inf` for every `ord=None` / `'fro'` / `ord=2` spelling.

## Note on float16

float16 is **not** affected: `linalg.vector_norm` accumulates low-precision inputs in
float32 (cf. #141953), which is enough headroom that no float16 input with a representable
norm can overflow the accumulator. float16 is correct for exactly the reason float32 and
float64 are not — there is no wider accumulation type in play for them. Passing float32
input through a float64 accumulation manually (`torch.linalg.norm(x.double()).float()`)
returns `3.0e19` correctly, confirming the reduction is where the range is lost.

### Versions

Collecting environment information...
PyTorch version: 2.14.0.dev20260809+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 16.0.6 (https://github.com/llvm/llvm-project.git 7cbf1a2591520c2491aa35339f227775f4d3adf6)
CMake version: version 4.1.0
Libc version: glibc-2.39

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-134-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No devices found.
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  384
On-line CPU(s) list:                     0-383
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      41%
CPU max MHz:                             3709.3569
CPU min MHz:                             1500.0000
BogoMIPS:                                4799.77
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc amd_ibpb_ret arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               6 MiB (192 instances)
L1i cache:                               6 MiB (192 instances)
L2 cache:                                192 MiB (192 instances)
L3 cache:                                768 MiB (24 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-95,192-287
NUMA node1 CPU(s):                       96-191,288-383
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] torch==2.14.0.dev20260809+cpu
[pip3] torchvision==0.29.0.dev20260810+cpu
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] torch                     2.14.0.dev20260809+cpu          pypi_0    pypi
[conda] torchvision               0.29.0.dev20260810+cpu          pypi_0    pypi

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano

### PR DESCRIPTION
**Pull request by dhruvildave235**
Fixes #193006

## Problem

`torch.linalg.norm` / `vector_norm` compute `sqrt(sum(x**2))` directly, so
overflow is determined by the intermediate `sum(x**2)`, not by the actual
norm. A float32 vector with `‖v‖ ≈ 1e20` — well inside float32's range
(max ≈ 3.4e38) — returns `inf`, because `sum(x**2) ≈ 1e40` overflows.
The gradient is worse: `backward()` evaluates `x / ‖x‖` as `x / inf` and
silently returns an all-zero gradient, with no `inf`/`nan` to signal
anything went wrong. `torch.nn.functional.normalize` inherits the same
bug since it's built on the same reduction.

## Fix

Rescales on the fly, LAPACK `dnrm2`-style, instead of squaring raw
elements. Implemented as an **online, associative accumulator** —
`(scale, ssq)` per element/lane, where `norm² == scale² · ssq` — rather
than a two-pass max-then-sum, so it drops directly into the existing
`reduce` / `combine` / `project` reduction framework:

reduce(acc, x):
ax = |x|
if ax == 0: return acc
if acc.scale < ax:
acc.ssq = 1 + acc.ssq * (acc.scale/ax)^2
acc.scale = ax
else:
acc.ssq += (ax/acc.scale)^2

combine(a, b):
# merges two partial (scale, ssq) accumulators — used for
# tree/parallel reduction and for merging SIMD lanes
...

project(acc):
return acc.scale * sqrt(acc.ssq)

Every intermediate term stays ≤ 1, so the reduction can't overflow even
when the final norm is well within the dtype's range. `scale == 0` is
the identity element, so this composes correctly with
`binary_kernel_reduce`'s parallel combine.

### Changes

- **`aten/src/ATen/native/SharedReduceOps.h`** — `NormTwoOps` now
  accumulates via a `NormTwoAccumulator<acc_t>{scale, ssq}` struct
  instead of a raw scalar sum. This is the general/fallback reduction
  path (used by CUDA, and by CPU whenever the vectorized fast path
  below doesn't apply).
- **`aten/src/ATen/native/cpu/ReduceOpsKernel.cpp`** — the vectorized
  last-dim fast path (`norm_two_reduce_step`, used for
  float32/float64/bfloat16 reducing the last dim — the path that
  actually executes for the bug's repro) reworked to track `(scale_vec,
  ssq_vec)` per SIMD lane, with lanes merged via the same combine rule
  at the end.
- `linalg._powsum` (`sum(|x|^p)`, no final root) shares `NormTwoOps` /
  `norm_two_reduce_step` for `p=2`, so it gets the same overflow fix for
  free — `project`/the final store return `scale² · ssq` instead of
  `scale · sqrt(ssq)` when the root shouldn't be applied.

`ord=2` matrix norm (spectral) goes through SVD and is a separate,
unaffected path (confirmed in the issue thread). `ord="fro"` calls
`vector_norm` directly and is fixed by this change. `ord=1/-1/inf/-inf`
use sum-of-absolute-values / max-abs, no squaring, unaffected.

## Testing

Verified the accumulator algorithm by hand against the issue's repro
case and a few edge cases (near-max-magnitude vectors, tiny/underflow-
prone vectors, zero vectors) — matches naive `sqrt(sum(x**2))`
everywhere that doesn't overflow, and stays finite where the current
code returns `inf`.

I have not been able to build PyTorch locally to compile-check this,
so opening as a **draft PR** for CI to validate.
```python
# see verify_norm.py in this PR / linked from the issue
```

**I have not been able to build PyTorch locally to compile-check this**,
so opening as a **draft PR** for CI to validate. Specific risk areas I'd
flag for a first review pass:

- `Vectorized<T>::blendv` / `operator!=` / `operator<` usage on
  `Vectorized<float>` / `Vectorized<double>` in the new
  `norm_two_reduce_step` — written against my reading of the header, not
  compiled.
- The recursive call to `norm_two_reduce_step` inside its own
  `BFloat16` specialization (unpacks to two float halves, then calls
  back into the primary float template) — want to confirm this resolves
  to the primary template as intended and not infinite-recurse into the
  specialization.
- Haven't located/audited the CUDA kernel registering `norm_stub` /
  `powsum_stub` to confirm it goes through `NormTwoOps` unmodified (in
  which case the `SharedReduceOps.h` change covers it) or has its own
  fast path needing separate treatment.
- Perf: the extra compare/select and divide per element in the
  vectorized path is real overhead vs. today's `acc += x*x`. Given this
  issue is tagged `low priority` / `module: edge cases`, correctness
  over overflow seems like the right tradeoff, but flagging for
  maintainer input rather than assuming.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @soulitzer @shadow-orange41 — building on the approach
@dhruvildave235 outlined in the issue thread.

### CODE REVIEW

**Comment by dhruvildave235** [comment]
@soulitzer this is a draft implementing the fix discussed in #193006 — flagging since CI workflows need maintainer approval to run on a first-time contribution. No rush, just didn't want it to sit unnoticed.

**Comment by dhruvildave235** [comment]
Bumping in case this fell through the cracks — still waiting on workflow
approval to get an actual CI signal on this. No urgency, just flagging.
